pulsar-broker/src/main/java/org/apache/pulsar/broker/delayed/bucket/ImmutableBucket.java

```diff
@@ -54,7 +65,28 @@ CompletableFuture<List<DelayedIndex>> asyncLoadNextBucketSnapshotEntry(boolean i
             final long bucketId = getAndUpdateBucketId();
             CompletableFuture<Integer> loadMetaDataFuture = new CompletableFuture<>();
             if (isRecover) {
-                // TODO Recover bucket snapshot
+                final long cutoffTime = cutoffTimeSupplier.get();
+                // Load Metadata of bucket snapshot
+                bucketSnapshotStorage.getBucketSnapshotMetadata(bucketId).thenAccept(snapshotMetadata -> {
+                    List<DelayedMessageIndexBucketSnapshotFormat.SnapshotSegmentMetadata> metadataList =
+                            snapshotMetadata.getMetadataListList();
+
+                    // Skip all already reach schedule time snapshot segments
+                    int nextSnapshotEntryIndex = 0;
+                    while (nextSnapshotEntryIndex < metadataList.size()
+                            && metadataList.get(nextSnapshotEntryIndex).getMaxScheduleTimestamp() <= cutoffTime) {
+                        nextSnapshotEntryIndex++;
+                    }
+
+                    this.setLastSegmentEntryId(metadataList.size());
+                    this.recoverDelayedIndexBitMapAndNumber(nextSnapshotEntryIndex, metadataList);
+
+                    int nextSegmentEntryId = nextSnapshotEntryIndex + 1;
+                    loadMetaDataFuture.complete(nextSegmentEntryId);
+                }).exceptionally(ex -> {
+                    loadMetaDataFuture.completeExceptionally(ex);
```

Comment:
It looks like we can simplify the logic to avoid creating the `loadMetaDataFuture`.
It will make the code easy to read.

Reply:
It seems that he still needs `loadMetaDataFuture`, but I improve the code, make the code easy to read